### PR TITLE
interpret: reduce usage of TypingEnv::fully_monomorphized

### PR DESCRIPTION
**Pull request by RalfJung**
r? @lcnr

### CODE REVIEW

**Comment by saethlin** [comment]
FYI `TypingEnv::fully_monomorphized` is faster than the alternatives because it just sets up fields instead of calling a query. I don't disagree with anything in this diff, but there may be a small perf hit that's justified by being more correct in implementation.

**Comment by RalfJung** [comment]
@bors try
@rust-timer queue

**Comment by RalfJung** [comment]
> justified by being more correct in implementation.

I guess the question is whether this is actually more correct. We know the DefId we are resolving here, and we know it has to be a monomorphic function. So it *should* be equivalent, right?

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -305,17 +306,18 @@ impl<'tcx, Prov: Provenance> ImmTy<'tcx, Prov> {
...
+        // Can use any typing env, since `Ordering` is always monomorphic.
         let ty = tcx.ty_ordering_enum(None);
         let layout =
             tcx.layout_of(ty::TypingEnv::fully_monomorphized().as_query_input(ty)).unwrap();
         Self::from_scalar(Scalar::from_i8(c as i8), layout)
     }
 
-    pub fn from_pair(a: Self, b: Self, tcx: TyCtxt<'tcx>) -> Self {
-        let layout = tcx
+    pub fn from_pair(a: Self, b: Self, cx: &(impl HasTypingEnv<'tcx> + HasTyCtxt<'tcx>)) -> Self {
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/interpret/operand.rs:316]: i guess the current assumption is that scalar pairs are always monomorphic?

**Review by lcnr — COMMENTED** [review]
r=me on the changes assuming that perf is green

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs:316]: This is currently only used for the return values of operations like AddWithOverflow, which is always monomorphic. But seems better to not implicit assume this, and the new version is actually more ergonomic to use than the old one...

**Comment by RalfJung** [comment]
Looks like the build worked except for uploading some metrics somewhere?
@bors try

**Comment by fmease** [comment]
Spurious, what the-. @bors retry

**Comment by RalfJung** [comment]
@bors retry rollup=always
this doesn't need its own dedicated bors slot

**Pull request merged**